lib/cloud-watch-event-listener.js

```diff
@@ -104,15 +104,104 @@ class CloudWatchEventListener extends events.EventEmitter {
     let region = body.region;
     let id = body.detail['instance-id'];
     let state = body.detail.state;
+    // Not a great name, but this variable is the time that the CloudWatch
+    // Event message was generated by the API, not the time at which we
+    // received it
+    let generated = new Date(body.time);
 
+    // We're logging this information to a database table so that we can
+    // see when these events happen
     try {
-      await this.state.logCloudWatchEvent({region, id, state});
+      await this.state.logCloudWatchEvent({region, id, state, received: generated});
     } catch (err) {
-      // We don't want to block things
+      // We don't want to block things, but let's bubble up the error for the
+      // time being
       this.monitor.reportError(err);
     }
 
-    if (state === 'pending' || state === 'running') {
+    let dbResponse;
+
+    let transaction = await this.state.beginTransaction();
+
+    try {
+      let x = await this.state.listInstances({region, id}, transaction);
+      assert(Array.isArray(x));
+      assert(x.length === 0 || x.length === 1);
+      if (x.length === 1) {
+        dbResponse = x[0];
+      }
+    } catch (err) {
+      await this.state.rollbackTransaction(transaction);
+      log.error(err, 'Error looking up state from database');
+    }
+
+    // We want to close this transaction if we didn't find anything in the database.
+    // We'll use a new transaction if we're looking up from the API, but if we got
+    // the information from the database, let's store it with the same transaction
+    if (!dbResponse) {
+      await this.state.commitTransaction(transaction);
+    }
+
+    if (dbResponse) {
+      // If there's a response from the database, we know that we already have
+      // this information in the database.  We'll use this information to
+      // figure out the immutable metadata
+
+      // Simple assertions, so may as well check for them.  If these failed,
+      // something would be foundationally wrong
+
+      assert(dbResponse.id === id);
+      assert(dbResponse.region === region);
+
+      let logInfo = {
+        workerType: dbResponse.workertype,
+        region: dbResponse.region,
+        az: dbResponse.az,
+        id: id,
+        instanceType: dbResponse.instancetype,
+        srid: dbResponse.srid,
+        imageId: dbResponse.imageId,
+        launched: dbResponse.launched,
+        state,
+        lastevent: generated,
+        metadataSource: 'db',
+      };
+      
+      if (dbResponse.lastevent < generated) {
+        try {
+          if (state === 'pending' || state === 'running') {
```

Comment:
So before we go too far into the discussion, I think it's worth noting that we don't get sent rebooting events.  Whether this is intentional, I'm not 100% sure.

```
ec2-manager-staging::DATABASE=> select distinct state from cloudwatchlog;
     state     
---------------
 pending
 running
 stopping
 shutting-down
 stopped
 terminated
(6 rows)
```
This is all instances in the ec2 account, not just staging ones.

I don't know if I would consider this to be a race condition, the instance *is* out of the pool during the time that it's rebooting, and even if `rebooting` events were sent, there would logically be a corresponding `running` event which would put it back into the calculations as running.